### PR TITLE
Annotate docker image label to deployment

### PR DESCRIPTION
**Pull request by Shudipta**

### CODE REVIEW

File: pkg/label-extractor/label_extractor.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	ac_log "github.com/appscode/go/log"
+
+	"sync"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+type LabelExtractor struct {
+	kubeClient kubernetes.Interface
+
```
Comment (diptadas) [inline pkg/label-extractor/label_extractor.go]: remove unnecessary newlines

File: pkg/label-extractor/label_extractor.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	return &LabelExtractor{
+		kubeClient: kubeClient,
+	}
+}
+
+func (l *LabelExtractor) Configure(enable bool) error {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	var err error
```
Comment (diptadas) [inline pkg/label-extractor/label_extractor.go]: wha't the use of error ?

File: pkg/label-extractor/util.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+const (
+	VerbUnchanged VerbType = ""
+	VerbCreated   VerbType = "created"
+	VerbPatched   VerbType = "patched"
+	VerbUpdated   VerbType = "updated"
+	VerbDeleted   VerbType = "deleted"
+)
+
+func PatchDeployment(
```
Comment (diptadas) [inline pkg/label-extractor/util.go]: use helpers from appscode/kutil

File: pkg/label-extractor/annotate_docker_labels.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+func (l *LabelExtractor) GetAllSecrets(sec []corev1.LocalObjectReference) []string {
+	var secs []string
+	for _, secName := range sec {
+		secs = append(secs, secName.Name)
+	}
+
+	return secs
+}
+
+func (l *LabelExtractor)Merge(mp1, mp2 map[string]string) map[string]string {
```
Comment (diptadas) [inline pkg/label-extractor/annotate_docker_labels.go]: use helper from kutil
https://github.com/appscode/kutil/blob/60ced7f055b34d512a84f65ca0a7d412ec3d6ef2/core/v1/kubernetes.go#L227

File: pkg/label_extractor/label_extractor.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	oldRes, ok := oldObj.(*v1beta1.Deployment)
+	if !ok {
+		return
+	}
+	newRes, ok := newObj.(*v1beta1.Deployment)
+	if !ok {
+		return
+	}
+
+	if oldRes.ResourceVersion == newRes.ResourceVersion {
```
Comment (tamalsaha) [inline pkg/label_extractor/label_extractor.go]: We should ignore if container images are same.

File: pkg/operator/operator.go [hunk 6]
```diff
@@ -158,6 +162,7 @@ func (op *Operator) Configure() error {
 func (op *Operator) setupWorkloadInformers() {
 	deploymentInformer := op.kubeInformerFactory.Apps().V1beta1().Deployments().Informer()
 	op.addEventHandlers(deploymentInformer, apps.SchemeGroupVersion.WithKind("Deployment"))
+	deploymentInformer.AddEventHandler(op.extractDockerLabel.ExtractDockerLabelHandler())
```
Comment (tamalsaha) [inline pkg/operator/operator.go]: This needs to support all workload types.

- Deployment
- ReplicaSet
- ReplicationContorller
- DaemonSet
- StatefulSet
- Job

**Comment by tamalsaha** [comment]
Use https://github.com/hashicorp/golang-lru to cache result for Docker image tag.

**Comment by tamalsaha** [comment]
Moved to https://github.com/kubepack/pack-server/pull/4

**Pull request closed without merging**